pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java

```diff
@@ -0,0 +1,199 @@
+import org.apache.pinot.segment.spi.creator.name.UploadedRealtimeSegmentNameGenerator;
+import org.apache.pinot.segment.spi.index.metadata.SegmentMetadataImpl;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.data.readers.RecordReader;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.roaringbitmap.RoaringBitmap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Minion task that compacts and merges multiple segments of an upsert table and uploads it back as one single
+ * segment. This helps in keeping the segment count in check and also prevents a lot of small segments created over
+ * time.
+ */
+public class UpsertCompactMergeTaskExecutor extends BaseMultipleSegmentsConversionExecutor {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpsertCompactMergeTaskExecutor.class);
+
+  public UpsertCompactMergeTaskExecutor(MinionConf minionConf) {
+    super(minionConf);
+  }
+
+  @Override
+  protected List<SegmentConversionResult> convert(PinotTaskConfig pinotTaskConfig, List<File> segmentDirs,
+      File workingDir)
+      throws Exception {
+    int numInputSegments = segmentDirs.size();
+    _eventObserver.notifyProgress(pinotTaskConfig, "Converting segments: " + numInputSegments);
+    String taskType = pinotTaskConfig.getTaskType();
+    Map<String, String> configs = pinotTaskConfig.getConfigs();
+    LOGGER.info("Starting task: {} with configs: {}", taskType, configs);
+    long startMillis = System.currentTimeMillis();
+
+    String tableNameWithType = configs.get(MinionConstants.TABLE_NAME_KEY);
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
+    Schema schema = getSchema(tableNameWithType);
+
+    SegmentProcessorConfig.Builder segmentProcessorConfigBuilder =
+        new SegmentProcessorConfig.Builder().setTableConfig(tableConfig).setSchema(schema);
+
+    // Progress observer
+    segmentProcessorConfigBuilder.setProgressObserver(p -> _eventObserver.notifyProgress(_pinotTaskConfig, p));
+
+    // get list of segment metadata
+    List<SegmentMetadataImpl> segmentMetadataList = segmentDirs.stream().map(x -> {
+      try {
+        return new SegmentMetadataImpl(x);
+      } catch (Exception e) {
+        throw new RuntimeException(String.format("Error fetching segment-metadata for segmentDir: %s", x), e);
+      }
+    }).collect(Collectors.toList());
+
+    // validate if partitionID is same for all small segments. Get partition id value for new segment.
+    int partitionID = getCommonPartitionIDForSegments(segmentMetadataList);
+
+    // get the max creation time of the small segments. This will be the index creation time for the new segment.
+    Optional<Long> maxCreationTimeOfMergingSegments = segmentMetadataList.stream().map(
+        SegmentMetadataImpl::getIndexCreationTime).reduce(Long::max);
+    if (maxCreationTimeOfMergingSegments.isEmpty()) {
```

Comment:
Just to confirm: `numInputSegments` is always greater than 0 right?

Reply:
Yes we have added the check in generator code to ensure `numInputSegments > 1`